awscli/customizations/assumerole.py

```diff
@@ -36,8 +38,41 @@ def inject_assume_role_provider_cache(session, **kwargs):
                   "assume-role cred provider cache.  Not configuring "
                   "JSONFileCache for assume-role.")
         return
-    provider = cred_chain.get_provider('assume-role')
-    provider.cache = JSONFileCache()
+    cred_chain.get_provider('assume-role').cache = JSONFileCache()
+    saml_provider = cred_chain.get_provider('assume-role-with-saml')
+    saml_provider.cache = JSONFileCache()
+    saml_provider.role_selector = role_selector
+
+
+def role_selector(role_arn, roles):
+    if role_arn:
+        chosen_roles = [r for r in roles if r['RoleArn']==role_arn]
+        if not chosen_roles:
+            raise ValueError(
+                'Your specified role %s does not exist in IdP roles list: %s'
+                % (role_arn, [r['RoleArn'] for r in roles]))
+        role = chosen_roles[0]
+    else:
+        # So end user has not yet configured a role_arn.
+        # If IdP provides more than one role, ask the user to choose one,
+        # otherwise just proceed.
+        if len(roles) > 1:
+            options = [(str(i), role) for i, role in enumerate(roles, start=1)]
+            prompt = "\n".join(
+                "[%s]: %s" % (i, role['RoleArn']) for i, role in options)
+            choice = raw_input(
+                "Please choose the role you would like to assume:\n"
+                "(You can preconfigure the role_arn in your profile)\n"
+                "%s\n"
+                "Selection: " % prompt)
+            role = dict(options).get(choice)
+            if not role:
+                raise ValueError('You selected an invalid role index.')
+        elif roles:
+            role = roles[0]
```

Comment:
The AWS profiles have always maintained the property that they correspond to a single set of permissions (i.e one credential pair, one role).

I don't think it's unreasonable that if you want to use a different role, you have to rerun `aws configure`.  This is also what AWS powershell does (maps one profile to one role).  Seems like another reason to stay consistent.

If we separate concerns, the only downside I see is you have to run `aws configure ...` to reconfigure a role if it changes.  I think being explicit about this also serves as an acknowledgement to the user that they understand what role they're picking.

Also, let's say a user *wants* to explicitly set a role but doesn't know the full role ARN.  If we leave this change as is, the user has to make some arbitrary CLI call, look at the RoleArns from the printed list, then copy/paste that into their AWS config file.  If this is exposed via `aws configure` they can easily set that up and select the correct role.